README
Statistics-Robust

Statistics::Robust is a perl implementation of a few of the methods described
in the books "Introduction to Robust Estimation and Hypothesis Testing" and
"Fundamentals of Modern Statistical Methods" by Rand Wilcox.  Many of the
functions were implemented with copious hints from the R functions provided by
the author on his website:

http://www-rcf.usc.edu/~rwilcox/


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Statistics::Robust

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Statistics-Robust

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Statistics-Robust

    CPAN Ratings
        http://cpanratings.perl.org/d/Statistics-Robust

    Search CPAN
        http://search.cpan.org/dist/Statistics-Robust


COPYRIGHT AND LICENCE

Copyright (C) 2008 Walter Szeliga

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Statistics-Robust

0.01    Thu July 30 2009
        - Original version uploaded to CPAN.



lib/Statistics/Robust.pm
package Statistics::Robust;

use warnings;
use strict;

use base 'Exporter';

sub
_min
{
 my($x,$y) = @_;

 if( $x <= $y )
 {
  return $x; 
 }
 else
 {
  return $y;
 }
}

sub
_sum
{
 my($x) = @_;

 my $sum = 0;
 foreach my $val (@$x)
 {
  $sum += $val;
 }

 return $sum;
}

1;

=head1 NAME

Statistics::Robust - A Collection of Robust Statistical Methods

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

use Statistics::Robust;

=head1 AUTHOR

Walter Szeliga, C<< <szeliga at colorado.edu> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-statistics-robust at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Statistics-Robust>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Statistics::Robust


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Statistics-Robust>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Statistics-Robust>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Statistics-Robust>

=item * Search CPAN

L<http://search.cpan.org/dist/Statistics-Robust>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2008 Walter Szeliga, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Statistics::Robust


lib/Statistics/Robust/Density.pm
package Statistics::Robust::Density;

use strict;
use warnings;
use Carp;

use Math::CDF qw(qnorm);
use Statistics::Robust::Location qw(mean);
use Statistics::Robust::Scale qw(variance);

use base 'Statistics::Robust';

our @EXPORT = ();
our @EXPORT_OK = (qw(
rdplot
akerd
));
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

# Expected Frequency Curve
sub
rdplot
{
 my($x,$fr) = @_;

 if ( not defined $fr )
 {
  $fr = 0.8;
 }

 my $n = scalar @$x;
 my @y = sort {$a <=> $b} @$x;
 my $mad = Statistics::Robust::Scale::MADN(\@y);
 my @rmd = ();

 for(my $i=0;$i<$n;$i++)
 {
  my $near = _near(\@y,$y[$i],$mad,$fr);
  $rmd[$i] = Statistics::Robust::_sum($near);
 } 
 if ( $mad != 0 )
 {
  for(my $i=0;$i<$n;$i++)
  {
   $rmd[$i] = $rmd[$i]/(2*$fr*$mad);
  }
 }
 for(my $i=0;$i<$n;$i++)
 {
  $rmd[$i] = $rmd[$i]/$n;
 }

 return \@y,\@rmd;
}

# Adaptive kernel density estimate
sub
akerd
{
 my($xx,$fr,$aval) = @_;

 # Set the variance calculation to use the un-biased estimate
 my $prior_value = 0;
 if ( defined $ENV{UNBIAS} )
 {
  $prior_value = $ENV{UNBIAS};
 }
 $ENV{UNBIAS} = 1; 

 if ( not defined $fr )
 {
  $fr = 0.8;
 }
 if ( not defined $aval )
 {
  $aval = 0.5;
 }

 my @x = sort {$a <=> $b} @$xx;

 # First, a measure of dispersion
 my $m = 0;
 $m = Statistics::Robust::Scale::MADN(\@x);
 if ($m == 0)
 {
  my($ql,$qu) = Statistics::Robust::Scale::idealf(\@x);
  $m = ($qu - $ql)/(qnorm(0.75) - qnorm(0.25));
 }
 if ( $m == 0 )
 {
  $m = sqrt(Statistics::Robust::Scale::winvar(\@x)/0.4129);
 }
 if ( $m == 0 )
 {
  carp "All measures of dispersion are equal to 0\n";
  return undef;
 }

 # Estimate the density using the Expected Frequency Curve 
 my(undef,$fhat) = rdplot(\@x);
 if ( $m > 0 )
 {
  for(my $i=0;$i<@$fhat;$i++)
  { 
   $fhat->[$i] = $fhat->[$i]/(2*$fr*$m);
  }
 }

 # Calculate the span
 my $n = scalar @x;
 my $sig = sqrt(variance(\@x));
 my($ql,$qu) = Statistics::Robust::Scale::idealf(\@x);
 my $iqr = ($qu-$ql)/1.34;
 my $A = Statistics::Robust::_min($sig,$iqr);
 if( $A == 0 )
 {
  $A = sqrt(Statistics::Robust::Scale::winvar(\@x))/0.64;
 }
 my $hval = 1.06*$A/($n**0.2);
 # See Silverman, 1986, pp. 47-48

 my @log_fhat = ();
 my @alam = ();
 for(my $i=0;$i<scalar @$fhat;$i++)
 {
  if ( $fhat > 0 )
  {
   $log_fhat[$i] = log($fhat->[$i]);
  }
 }
 my $gm = exp(mean(\@log_fhat));
 for(my $i=0;$i<scalar @$fhat;$i++)
 {
  $alam[$i] = ($fhat->[$i]/$gm)**(0.0-$aval);
 }

 my @dhat = ();
 my @temp = ();
 for(my $j=0;$j<$n;$j++)
 { 
  # Calculate t
  for(my $i=0;$i<$n;$i++)
  {
   $temp[$i] = ($x[$j]-$x[$i])/($hval*$alam[$i]);
  }
 
  # Calculate the Epanechnikov kernel
  for(my $i=0;$i<$n;$i++)
  {
   my $epan = 0;
   if ( abs($temp[$i]) < sqrt(5) )
   {
    $epan = 0.75*(1-0.2*$temp[$i]**2)/sqrt(5);
   }
   else
   {
    $epan = 0.0;
   }
   $temp[$i] = ($epan/($alam[$i]*$hval));
  }
  $dhat[$j] = mean(\@temp) + 0.0;
 }

 # Principal of least surprise
 $ENV{UNBIAS} = $prior_value; 
 
 return \@x,\@dhat;
}

# Determine which values in $x are near $pt
sub
_near
{
 my($x,$pt,$m,$fr) = @_;

 if ( not defined $fr )
 {
  $fr = 1;
 }
 if ( $m == 0 )
 {
  my($ql,$qu) = idealf($x);
  $m = ($qu-$ql)/(qnorm(0.75) - qnorm(0.25));
 }
 if ( $m == 0 )
 {
  $m = $m = sqrt(winvar($x)/0.4129);
 }
 if ( $m == 0 )
 {
  carp "All measures of dispersion are equal to 0\n";
 }

 my @dflag = ();
 for(my $i=0;$i<scalar @$x;$i++)
 {
  my $dis = abs($x->[$i]-$pt);
  push(@dflag, ($dis <= $fr*$m));
 }

 return \@dflag;
}

1;

=head1 NAME

Statistics::Robust::Density - Robust Probability Density Estimators

=head1 SYNOPSIS

 my @x = (1,4,5,3,7,2,4);
 
 my($pts,$akerd) = akerd(\@x);

=head1 FUNCTIONS

=head2 rdplot

 my($pts,$rdplot) = rdplot(\@x);

 Return the expected frequency curve for the data in \@x.

=head2 akerd

 my($pts,$akerd) = akerd(\@x);

 Return the adaptive kernel density estimate for the data in \@x.

=head1 AUTHOR

Walter Szeliga C<< <szeliga@colorado.edu> >>

=cut


lib/Statistics/Robust/Location.pm
package Statistics::Robust::Location;

use strict;
use warnings;
use Carp;

use Math::CDF qw(pbeta);
use POSIX qw(floor);

use base 'Statistics::Robust';

our @EXPORT = ();
our @EXPORT_OK = (qw(
median
win
tmean
mean
hd
));
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

# Implement the median to avoid needing Statistics::Basic
# which has silly Number::Format requirements (Version 1.6004)
sub
median
{
 my($x) = @_;

 my @y = sort {$a <=> $b} @$x;
 my $length = scalar @y;

 my $median = undef;
 if ( $length % 2 ) 
 {
  my $lower_half = int (($length-1)/2);
  my $upper_half = int  ($length/2);

  $median = ($y[$lower_half] + $y[$upper_half])/2;
 } 
 else
 {
  my $half = int ($length/2);
  $median = $y[$half];  
 }

 return $median; 
}


# The gamma Winsorized mean
sub
win
{
 my($x,$gamma) = @_;

 if ( not defined $gamma )
 {
  $gamma = 0.2;
 } 

 my @y = sort {$a <=> $b} @$x;
 my $n = scalar @y;
 my $ibot = floor($gamma*$n) + 1;
 my $itop = $n - $ibot + 1;
 my $xbot = $y[$ibot];
 my $xtop = $y[$itop];

 for(my $i=0;$i<@y;$i++)
 {
  if ( $y[$i] <= $xbot )
  {
   $y[$i] = $xbot;
  }
  if ( $y[$i] >= $xtop )
  {
   $y[$i] = $xtop;
  }
 }
 
 my $win = mean(\@y);
 return $win;
}

# The trimmed mean
sub
tmean
{
 my($x,$tr) = @_;

 if ( not defined $tr )
 {
  $tr = 0.2;
 }

 my @y = sort {$a <=> $b} @$x;
 my $n = scalar @y;
 my $g = floor($tr*$n);

 my $tmean = 0.0; 
 for(my $i=$g;$i<($n-$g);$i++)
 {
  $tmean += $y[$i];
 } 
 $tmean = $tmean/($n - 2*$g);

 return $tmean; 
}

# The Harrel-Davis Estimator for quantiles
sub
hd
{
 my($x,$q) = @_;
 
 if ( not defined $q )
 {
  $q = 0.5;
 }
 my $n = scalar @$x;
 my $m1 = ($n+1)*$q;
 my $m2 = ($n+1)*(1-$q);
 my @y = sort  {$a <=> $b} @$x;

 my @wy = ();
 for(my $i=1;$i<=$n;$i++)
 {
  my $w = pbeta(($i/$n),$m1,$m2) - pbeta(($i-1)/$n, $m1,$m2);
  $wy[$i-1] = $w*$y[$i-1]; 
 }

 my $hd = Statistics::Robust::_sum(\@wy);

 return $hd;
}

sub
mean
{
 my($x) = @_;

 my $n = scalar @$x;
 if ( $n == 0 ) { return undef;}
 my $sum = Statistics::Robust::_sum($x);
 my $mean = $sum/$n;

 return $mean; 
}

1;

=head1 NAME

Statistics::Robust::Location - Robust Location Estimators

=head1 SYNOPSIS

 my @x = (1,4,5,3,7,2,4);

 my($tmean) = tmean(\@x);

 my($win) = win(\@x);

=head1 FUNCTIONS

=head2 win

 my($win) = win(\@x,$gamma);

 Return the gamma Winsorized mean.  If gamma is not specified, it defaults to 0.2.

=head2 tmean

 my($tmean) = tmean(\@x,$tr);

 Return the trimmed mean.  If $tr is not specified, it defaults to 0.2.

=head2 mean

 my($mean) = mean(\@x);

 Although not a robust measure of location, it is useful for comparison as well as used internally in the computation
 of some robust methods. This returns the mean of the array @x.

=head2 hd

 my($est) = hd(\@x, $q)

 The Harell-Davis Estimator for the quantile $q.

=head2 median

 my($median) = median(\@x);

 Return the median

=head1 AUTHOR

Walter Szeliga C<< <szeliga@colorado.edu> >>

=cut


lib/Statistics/Robust/Bootstrap.pm
package Statistics::Robust::Bootstrap;

use strict;
use warnings;
use Carp;

use POSIX qw(floor);
use Math::Round;
use Math::Random;

use base 'Statistics::Robust';

our @EXPORT = ();
our @EXPORT_OK = (qw(
onesample
sample_wr
sample_wor
));
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

# The percentile bootstrap
sub
onesample
{
 my($x,$func,$alpha,$n) = @_;

 my @b_vec = ();
 for(my $i=0;$i<$n;$i++)
 {
  my($y) = sample_wr($x);
  my $val = $func->($y);
  push(@b_vec,$val);
 }

 @b_vec = sort {$a <=> $b} @b_vec;
 my $low = round(($alpha/2)*$n);
 my $up = $n-$low;
 $low++;
 
 my $cl = $b_vec[$low];
 my $ch = $b_vec[$up];

 return $cl,$ch;
}

# Sampling with replacement

sub
sample_wr
{
 my($x) = @_;

 my $n = scalar @$x;
 my @result = ();
 my(@j) = random_uniform_integer($n,0,($n-1));
 for(my $i=0;$i<$n;$i++)
 {
  $result[$i] = $x->[$j[$i]];
 }

 return \@result;
}

# Sampling without replacement
# Algorithm 3.4.2S from Knuth's Seminumeric Algorithms

sub
sample_wor
{
 my($x,$n) = @_;
 
 my $N = scalar @$x; # Population Size
 # $n is the sample size

 my @samples = (); # Indices of selected items
 my $t = 0; # total input records dealt with
 my $m = 0; # number of items selected so far

 while ( $m < $n )
 {
  my $u = random_uniform();

  if ( ($N - $t)*$u >= ($n-$m) )
  {
   $t++;
  }
  else
  {
   $samples[$m] = $x->[$t];
   $t++;
   $m++;
  }
 }
 
 return \@samples;
}

1;

=head1 NAME

Statistics::Robust::Bootstrap - Bootstrap Estimation

=head1 SYNOPSIS

 my @x = (1,4,5,3,7,2,4);

 my($low,$high) = onesample(\@x, \&func, $alpha, $n);

 my($wr) = sample_wr(\@x);
 
 my($wor) = sample_wor(\@x,$n);

=head2 onesample

 my($low,$high) = onesample(\@x, \&mean, $alpha, $n);

Compute a bootstrap estimate of the alpha confidence interval for a measure of
location func using a percentile bootstrap method.

=head2 sample_wr

 my($resamp) = sample_wr(\@x);

Sample with replacement from a vector

=head2 sample_wor

 my($resamp) = sample_wor(\@x,$n);

Sample $n values without replacement from a vector

=head1 AUTHOR

Walter Szeliga C<< <szeliga@colorado.edu> >>

=cut


lib/Statistics/Robust/Scale.pm
package Statistics::Robust::Scale;

use strict;
use warnings;
use Carp;

use POSIX qw(floor);
use Math::CDF qw(qnorm pbeta);
use Math::Round qw(round_even);
use Statistics::Robust::Location qw(median);

use base qw(Statistics::Robust);

our @EXPORT = ();
our @EXPORT_OK = (qw(
variance
MAD
MADN
idealf
winvar
trimvar
msmedse
pbvar
));
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

# We need to implement variance since Statistics::Basic
# currently is messing up the unbiased variance implementation.
sub
variance
{
 my($x) = @_;

 my $length = scalar @$x;
 my $sum = Statistics::Robust::_sum($x);
 my $mean = $sum/$length;

 my $var = 0;
 for(my $i=0;$i<@$x;$i++)
 {
  $var += ($mean-$x->[$i])**2;
 }

 return $var/($length-1);
}

# The Median Absolute Deviation
sub
MAD
{
 my($x) = @_;

 my @ad = ();

 my($median) = Statistics::Robust::Location::median($x);

 foreach my $xi (@$x)
 {
  my($adi) = abs($xi - $median);
  push(@ad,$adi);
 }

 my($mad) = Statistics::Robust::Location::median(\@ad) + 0.0;


 return $mad;
}

# The rescaled Median Absolute Deviation
sub
MADN
{
 my($x) = @_;
 
 my $mad = MAD($x);
 $mad /= qnorm(0.75);

 return $mad;
}

sub
idealf
{
#
# Compute the ideal fourths for data in x
# and return the lower and upper quartiles
#
 my($x) = @_;

 my $n = scalar @$x;

 my $j = floor($n/4 + 5/12) - 1;
 my @y = sort {$a <=> $b} @$x;
 my $g = ($n/4) - $j + (5/12) - 1;
 my $ql = (1-$g)*$y[$j] + $g*$y[$j+1];
 my $k = $n - $j - 1;
 my $qu = (1-$g)*$y[$k] + $g*$y[$k-1];

 return ($ql,$qu); 
}

sub
winvar
{
 my($x,$tr) = @_;

 if ( not defined $tr )
 {
  $tr = 0.2;
 }
 my @y = sort {$a <=> $b} @$x;

 my $n = scalar @$x;
 my $ibot = floor($tr*$n);# + 1;
 my $itop = $n - $ibot;# + 1;
 my $xbot = $y[$ibot]; 
 my $xtop = $y[$itop]; 
 for(my $i=0;$i<$n;$i++)
 {
  if ( $y[$i] <= $xbot )
  {
   $y[$i] = $xbot;
  }
  if ( $y[$i] >= $xtop )
  {
   $y[$i] = $xtop;
  }
 }
 my $winvar = variance(\@y);

 return $winvar;
}

# The variance of the trimmed mean
sub
trimvar
{
 my($x,$tr) = @_;

 if ( not defined $tr )
 {
  $tr = 0.2;
 }

 my $n = scalar @$x;
 my $winvar = winvar($x,$tr);
 my $trimvar = $winvar/((1-2*$tr)**2*$n);

 return $trimvar;
}

# Given an array, estimate the standard error of the sample median from pg. 65
# of Wilcox, "Introduction to Robust Estimation and Hypothesis Testing", 2005

sub
msmedse
{
 my($x) = @_;
 
 my @sorted = sort {$a <=> $b} @$x;
 unshift @sorted, 0.0;
 my $n = @sorted -1;
 my $av = round_even(($n+1)/2.0 - 2.575829*sqrt($n/4.0));
 if ( $av == 0 )
 {
  $av = 1.0; 
 }
 my $top = ($n - $av + 1);
 my $sqse = (($sorted[$top] - $sorted[$av])/(2*2.575829))**2;
 $sqse = sqrt($sqse);

 return $sqse;
}

# The percentage bend midvariance
sub
pbvar
{
 my($x,$beta) = @_;

 if ( not defined $beta )
 {
  $beta = 0.2;
 }
 my $pbvar = 0;
 my $n = scalar @$x;
 my $median = Statistics::Robust::Location::median($x) + 0.0;

 my @w = ();
 for(my $i=0;$i<@$x;$i++)
 {
  $w[$i] = ($x->[$i] - $median);
 }

 my @sorted = sort {abs($a) <=> abs($b)} @w;
 my $m = floor((1-$beta)*$n + 0.5); 
 my $omega = $sorted[$m];

 if ( $omega > 0 )
 {
  my @z=0;
  my $np = 0;

  for(my $i=0;$i<@w;$i++)
  {
   my $y = $w[$i]/$omega; 
   if ( $y >= 1.0 )
   {
    $y = 1.0;
   }
   elsif ( $y <= -1 )
   {
    $y = -1.0;
   }
   else
   {
    $np++;
   }
   $z[$i] = $y**2;
  }
  
  $pbvar = $n*$omega**2*Statistics::Robust::_sum(\@z)/($np**2);
 }

 return $pbvar;
}

1;

=head1 NAME

Statistics::Robust::Scale - Robust Measures of Scale

=head1 SYNOPSIS

 my @x = (1,4,5,3,7,2,4);

 my($mad) = MAD(\@x);
 my($madn) = MADN(\@x);
 my($ql,qu) = idealf(\@x);
 my($winvar) = winvar(\@x);

=head1 FUNCTIONS

=head2 MAD

 my($mad) = MAD(\@x);

 Return the non-normalized Median Absolute Deviation.

=head2 MADN

 my($madn) = MADN(\@x);

 Return the Median Absolute Deviation normalized by the 0.75 quartile of the normal distribution.

=head2 idealf

 my($ql,$qu) = idealf(\@x); 

 Return the Ideal Fourths estimate of the lower and upper quartiles (in that order).

=head2 winvar

 my($winvar) = winvar(\@x,$tr);

 Return the Winsorized variance.  If the amount of trimming ($tr) is not specified, it defaults to 0.2.

=head2 trimvar

 my($trimvar) = trimvar(\@x,$tr);

 Return the variance for the trimmed mean with $tr trimming.  If the amount of trimming is not specified,
 it defaults to 0.2.

=head2 variance

 my($var) = variance(\@x);

 The unbiased sample variance.

=head2 msmedse

 my($mse) = msmedse(\@x);

 An estimate of the standard error of the median.

=head2 pbvar

 my($pb) = pbvar(\@x, $beta);

 The percentage-bend midvariance

=head1 AUTHOR

Walter Szeliga C<< <szeliga@colorado.edu> >>

=cut
